### PR TITLE
Proposal: server error middleware

### LINKED ISSUE #742: Custom `internal server error` messages
Sapper will currently bail with `<pre>Internal server error</pre>` when errors are thrown while rendering error pages.

This was done to prevent the `node` process from hanging in an infinite loop, but we should probably look into handling it more gracefully.

At the very least, we could support customizing the error with a static `template-internal-error.html` akin to `template.html` when nothing else can be done.

Thoughts?

### LINKED ISSUE #812: Proper error handling for server routes instead of sending all errors as a text response
**Is your feature request related to a problem? Please describe.**

All errors originating in server routes are currently sent as a text response by Sapper and there is no way to prevent this. This might cause to leak some information which should not happen on production. It also requires somehow boilerplate-ish code in every server route if you need to either log the error or to format it in some other way (e.g. formulating a JSON response). 

**Describe the solution you'd like**

An option to provide a function to handle server route errors in a centralised way.

**Describe alternatives you've considered**

Injecting express/polka middleware before or after Sapper's cannot solve this problem, because Sapper handles server routing internally in its middleware, catches and handles all errors inside.

**How important is this feature to you?**

I believe it's critical for any real-world production deployment to be able to control what happens with errors. You usually need to log them and you need to control the output sent to the browser. At least, it is critical to me.

**Additional context**

I created a patch for this and will submit a pull request in a moment.

### LINKED ISSUE #955: Allow server side redirect on 404 / 500 error.
**Is your feature request related to a problem? Please describe.**
I can't figure out how to redirect to the index page on a `404 error` without using client side routing after `onMount`. 

**Describe the solution you'd like**
A way for the server to detect a status code and redirect to the desired route . 

**Describe alternatives you've considered**
Perhaps implementing some sort of `preload`  `this.redirect` in the `_error.svelte` template that also has access to the status code so the redirect can be based on the status code.

**How important is this feature to you?**
I think for certain apps, i.e. a timesheet app for logging employee time, having all 404 routes lead to the index (which is the timesheet page) leads to a cleaner feeling app. On some apps, going to a 404 and having to click out of it can feel clunky or like something is broken (which in a way is), but redirecting to main page makes the app feel more polished.

### LINKED ISSUE #993: sapper.middleware catches all server errors, would rather pass through
**Is your feature request related to a problem? Please describe.**
I have a error handler in my express app that sends server errors to Sentry and does some custom error handling. Sapper's error handler catches all server errors and doesn't pass them through to my default.

**Describe the solution you'd like**
To pass through sapper server route errors to pass through to my error handling middleware

**Describe alternatives you've considered**
None

**How important is this feature to you?**
Error handling is muy importante

**Additional context**
No extra context

### PR DESCRIPTION
**Pull request by koenkivits**
fixes #742, fixes #812, fixes #955, fixes #993, fixes #960 

_First off: I implemented this before I saw https://github.com/sveltejs/sapper/pull/1512#issuecomment-729748586. I'm not sure if it has any chance of getting merged, but I still wanted to share it. I need it for myself anyway, because I need more flexible error handling and I couldn't wait for Kit right now. :)_

This PR adds support for custom error middleware. It allows you to define an error handler in `src/routes/_error.{js,ts}` like so:

```js
export default function(err, req, res, next) {
    handleError(err);

    next(); // fall back to Sapper error handling

    // or:
    next(otherErr); // fall back to Sapper error handling with different error
                    // (allows you to control what error is rendered)

    // ... or don't call next(). Sapper handling will not be invoked.
}
```

It also centralizes all Sapper handling. This means:

* _all_ errors that are logged or rendered by Sapper's error handling have their stack trace adjusted for the source map (this was only the case for page errors)
* server routes that throw or call `next(err)` also render a full error page when possible

... and it makes the default error handling respect `err.status` (and `err.statusCode`), like both Express and Polka.

This PR solves the same problem as #1512, but:

* it doesn't introduce a 'new' concept. The notion of `(err, req, res, next) => void` middleware already exists in Express/Polka and should be familiar to most users. This makes it easier to explain and makes it possible to reuse existing error handlers.
* it also improves the _default_ Sapper error handling (or at least I think so)
* no additional config (unless you consider a `src/routes/_error` file to be config :slightly_smiling_face:)

### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] _Ideally, include a test that fails without this PR but passes with it._ -> Not yet, but I'd be more than happy to add these if this PR has any change of getting merged. [Update]: the tests run successfully on my machine, but I see CI fails. I can have a look at that if necessary

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`

### CODE REVIEW

File: runtime/src/server/middleware/get_page_renderer.ts [hunk 1]
```diff
@@ -0,0 +1,402 @@
+import { writable } from 'svelte/store';
```
Comment (benmccann) [inline runtime/src/server/middleware/get_page_renderer.ts:1]: I think renaming this file as part of this PR makes the diff too hard to follow

**Comment by benmccann** [comment]
I didn't look too closely at this. A lot of the changes here seem generally agreeable. However, we're working on a successor to Sapper called [SvelteKit](https://svelte.dev/blog/whats-the-deal-with-sveltekit) and I'd rather any major changes go there given our limited review time. The code for it is still private at the moment while we work on getting things in a more stable state, but hopefully we'll have more to share before too long.

File: runtime/src/server/middleware/get_page_renderer.ts (hunk 1 above)
Comment (koenkivits) [reply runtime/src/server/middleware/get_page_renderer.ts]: You're right! I moved `get_page_renderer` back into `get_page_handler` to make the diff more readable (which probably also reduces the risk of merge conflicts)

**Comment by koenkivits** [comment]
@benmccann I understand, thanks for looking at it anyway! I know about SvelteKit (and am looking forward to it!), but I need this functionality soon for myself and SvelteKit has no public timeline yet. Good to know more details are coming soon. :)

**Comment by benmccann** [comment]
I'm going to close this given that we're only adding major new features in SvelteKit currently and making only small bug fixes in Sapper. Feel free to propose again in the SvelteKit repo once it's released

**Comment by benmccann** [comment]
Thanks again for this! There's a few merge conflicts. SvelteKit is public now, so maybe better to just focus on making sure SvelteKit works as desired?

**Comment by koenkivits** [comment]
@benmccann I haven't had the time to look at SvelteKit in detail yet, but I think you're right. I'll close this PR for now and will hopefully adapt it for SvelteKit if I have some spare time. :)

**Comment by benmccann** [comment]
Thanks!

**Comment by bamadesigner** [comment]
I came up with a simple fix for anyone who is interested/needs something to hold you over until SvelteKit.

I added the status code to the response sent during `handle_page()`, so when its passed back through middleware you can check response.statusCode inside the middleware function, e.g. `if (500 === response.statusCode)` equals true when a 500 error has been thrown.

I decided to fork sapper for myself since they might not do any more work here anyway. This is the only change I've made to my fork. Feel free to copy.

https://github.com/bamadesigner/sapper/pull/1/files
